File: api/fetch-shopify.py
import hashlib

from api._lib.blob_io import save_df
from api._lib.core.category_extraction import preferred_category_column
from api._lib.core.feed_processor import count_products
from api._lib.core.shopify_fetcher import fetch_shopify_products, shopify_keep_map
from api._lib.handler_utils import BaseHandler, column_meta


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            store_url = body.get("storeUrl", "").strip()
            if not store_url:
                return self._send_json(400, {"error": "storeUrl is required"})

            df = fetch_shopify_products(store_url)
            keep_map = shopify_keep_map(df)
            raw_url = save_df(df, "shopify_raw")
            file_hash = hashlib.md5(store_url.encode()).hexdigest()
            display_name = store_url.replace("https://", "").replace("http://", "").rstrip("/")

            self._send_json(200, {
                "columns": column_meta(df, keep_map),
                "productCount": count_products(df),
                "fileName": display_name,
                "sheetName": None,
                "sheets": None,
                "rawDfBlobUrl": raw_url,
                "fileHash": file_hash,
                "catSrcCol": preferred_category_column(df),
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/fetch-shopify.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import crypto from 'node:crypto'
import { sendError } from '../api/_lib/handlerUtils.js'
import { fetchShopifyProducts, shopifyKeepMap } from '../api/_lib/core/shopifyFetcher.js'
import { countProducts } from '../api/_lib/core/feedProcessor.js'
import { preferredCategoryColumn } from '../api/_lib/core/categoryExtraction.js'
import { columnMeta } from '../api/_lib/handlerUtils.js'
import { getOrCreateSession, saveRows } from '../api/_lib/sessionStore.js'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { storeUrl } = req.body as { storeUrl: string }
    if (!storeUrl?.trim()) return res.status(400).json({ error: 'storeUrl is required' })

    const rows = await fetchShopifyProducts(storeUrl.trim())
    const cols = Object.keys(rows[0] ?? {})
    const keepMap = shopifyKeepMap(cols)
    const fileHash = crypto.createHash('md5').update(storeUrl).digest('hex')
    const displayName = storeUrl.replace(/https?:\/\//, '').replace(/\/$/, '')

    const sessionId = await getOrCreateSession(fileHash, displayName)
    const rawRef = await saveRows(sessionId, 'raw', rows)

    res.status(200).json({
      rawDfBlobUrl: rawRef,
      sessionId,
      columns: columnMeta(rows, keepMap),
      productCount: countProducts(rows),
      fileName: displayName,
      sheetName: null,
      sheets: null,
      fileHash,
      catSrcCol: preferredCategoryColumn(cols),
    })
  } catch (err) {
    sendError(res, err)
  }
}
